### ISSUE TITLE
Rainier Power Host Interface - GetAlertStatus PLDM OEM cmd

### ISSUE THREAD

**Issue by derekhoward55** [ibm-openbmc/dev#890]
Implement GetAlertStatus PLDM OEM cmd to report enclosure ids of system, alert status and flags, and config ID.  Enclosure ID, Alert status, flags, and config ID can be defaulted to typical values, full support for filling these in based on alert interface will be done in a different issue. 

Eventually (with future issue) config ID can come from mapping of property "IM" under the "com.ibm.ipzvpd.vsys" (names could change) on the "/xyz/openbmc_project/inventory/system" D-Bus object in the inventory.
 
Command definition is below.  The document that contains the formatted definition and other pertinent information can be shared per request.

Type     | Request Data
--         | --
uint8    | formatVersion.	Version of the command/response format. 0x00 for this format.
---------------------------------------------------------------------------------

Type    | Response Data
--        | --
enum8 | completionCode.  value: { PLDM_BASE_CODES, UNSUPPORTED_FORMAT_VERSION = 0x81 }
uint32 | Enclosure ID, Alert Status, Flags, Config ID.  _(rack entry)_
uint32 | Enclosure ID, Alert Status, Flags, Config ID. _(pri cec node)_
uint32 | Enclosure ID, Alert Status, Flags, Config ID. _(additional slave node)_
uint32 | Enclosure ID, Alert Status, Flags, Config ID. _(additional slave node)_
...         | ...                  | ...                 | ...       | ...              | ...

For this commit, the default response data should be:
00 (Success)
FF 00 00 30
00 00 80 30

Implementing the cmd should be similar to the other pldm cmds, such as what was done for GetFRURecordTable:
https://github.com/ibm-openbmc/dev/issues/1044
https://github.com/ibm-openbmc/dev/issues/1195

**Comment by rfrandse** [comment]
refresh

**Comment by rfrandse** [comment]
refresh

**Comment by derekhoward55** [comment]
This work item is not dependent on Rainier hardware nor PHYP.  This new command can be sent on any type of system, and the tool that is used to test the other pldm cmds, such as the ones in the issues referenced above, could be used to test this code as well, thanks.

**Comment by bradbishop** [comment]
Hi @derekhoward55 just curious how the response data ends up getting used?

**Comment by derekhoward55** [comment]
Hi, sure @bradbishop , it allows PHYP&OSes to associate enclosure info and which enclosures have problems such as impending power failures due to power/thermal issues, powered on status, which enclosure is the primary cec node, changes to power/thermal frus or enclosures due to actions such as hot-plug/concurrent maintenance, UPS and early power-off warning status, and a system config ID specifically for IBMi mapping.  
What PHYP and OSes do with the information is pretty wide-ranging and I'd probably need to pull the various areas in to give a more detailed answer, but at a high level for the UPS/EPOW status the partitions would start to save off data and do an orderly shutdown before power is totally lost or the system power is pulled to prevent hw damage due to thermal concerns, to keep their view of the system and what they portray in step with reality after frus/enclosures are removed/added, etc.

**Comment by bradbishop** [comment]
Thanks @derekhoward55.  No need for more detail, thanks.  Just to make sure I have it right - the data flow is from the BMC to the system firmware (e.g. phyp) and it, not the BMC, is taking action on the response data (in other words the bmc is filling out the response) ?

**Comment by derekhoward55** [comment]
Correct.  BMC would have taken any action needed independently.  This is to allow phyp/os to do what they need to do based on the data (filled out by bmc).

**Comment by lxwinspur** [comment]
@derekhoward55 
Through my understanding of context, I'm a little confused.
Currently, for this task, we can only en/decode for the GetAlertStatus PLDM OEM command via the above spec doc, right?
Or is this epic?
If yes, have you created the related task?
thanks.

**Comment by derekhoward55** [comment]
Hi @lxwinspur, I'm sorry I don't quite understand the question.  The work for this story is to implement the GetAlertStatus PLDM OEM cmd on BMC.  The cmd can just return the hardcoded data I have listed above, copied here:

00 (Success)
FF 00 00 30
00 00 80 30

At a later time we'll update the response data to get the data from the appropriate source (https://github.com/ibm-openbmc/dev/issues/1863).  For now we just want to return the above response data to quickly allow the Host to be able to send the cmd.

**Comment by lxwinspur** [comment]
Thanks, @derekhoward55 
So basically this task involves:
* Implement decode/encode function for GetAlertStatus
  - Requester: 
    - encode_get_alert_status_req
    - decode_get_alert_status_resp
  - Responder:
    - encode_get_alert_status_resp
    - decode_get_alert_status_req
  - Add Unit Tests.
* Implement GetAlertStatus in oem/ibm/libpldmresponder.
* pldmtool: implement GetAlertStatus.

And the cmd can just return the hardcoded:
00 (Success)
FF 00 00 30
00 00 80 30

**Comment by lxwinspur** [comment]
* Implement decode/encode function for GetAlertStatus
  - Requester: 
    - https://gerrit.openbmc-project.xyz/c/openbmc/pldm/+/30226
  - Responder:
    - https://gerrit.openbmc-project.xyz/c/openbmc/pldm/+/30227
* Implement GetAlertStatus in oem/ibm/libpldmresponder
    - https://gerrit.openbmc-project.xyz/c/openbmc/pldm/+/30228
* pldmtool: implement GetAlertStatus command
    - https://gerrit.openbmc-project.xyz/c/openbmc/pldm/+/30233

**Comment by derekhoward55** [comment]
Wow, thanks for the quick turnaround, amazing!  I'll try to get up to speed on pldm and review them, hopefully Lei and Deepak can review them as well with their expertise.

Thanks!

**Comment by derekhoward55** [comment]
fyi the original estimate in this story was for all the meetings between bmc&phyp to re-design multiple legacy commands into just this one for the bmc/host interface over the last couple of months.  I then updated this story description for to include just the 1 GetAlertStatus command.  I understand why that estimate was confusing.

I reduced the estimate for this story to reflect just the GetAlertStatus hardcoded cmd.  That is all that is needed for this story.  Hopefully that clears it up.  Thanks!
